pkg/messaging/mqtt/pubsub_test.go

```diff
@@ -0,0 +1,599 @@
+			clientID: "clientid5",
+			err:      mqtt_pubsub.ErrFailed,
+		},
+		{
+			desc:     "Unsubscribe from a topic with an ID with failing handler",
+			topic:    fmt.Sprintf("%s%s.%s", topic, "2", subtopic),
+			clientID: "clientid5",
+			err:      mqtt_pubsub.ErrFailed,
+		},
+	}
+	for _, tc := range cases {
+		err := pubsub.Unsubscribe(tc.clientID, tc.topic)
+
+		switch tc.err {
+		case nil:
+			if tc.err != nil {
+				t.Errorf("%s: expected an error: %s, did not receive any", tc.desc, tc.err)
+			}
+		default:
+			switch tc.err {
+			case nil:
+				t.Errorf("%s: got unexpected error: %s", tc.desc, err)
+			default:
+				assert.Equal(t, err.Error(), err.Error(), fmt.Sprintf("%s: expected: %s, but got: %s", tc.desc, err, tc.err))
+			}
+		}
+	}
+}
+
+// Tests both publisher and subscriber
+func TestPubSub(t *testing.T) {
+	client, err := newClient(address, "", 30*time.Second)
+	t.Cleanup(func() {
+		defer client.Unsubscribe()
+		defer client.Disconnect(5)
+	})
+	require.Nil(t, err, fmt.Sprintf("got unexpected error while creating client: %s", err))
+
+	cases := []struct {
+		desc     string
+		topic    string
+		clientID string
+		err      error
+		handler  messaging.MessageHandler
+	}{
+		{
+			desc:     "Subscribe to a topic with an ID",
+			topic:    topic,
+			clientID: "clientid7",
+			err:      nil,
+			handler:  handler{false},
+		},
+		{
+			desc:     "Subscribe to the same topic with a different ID",
+			topic:    topic,
+			clientID: "clientid8",
+			err:      nil,
+			handler:  handler{false},
+		},
+		{
+			desc:     "Subscribe to a topic with a subtopic with an ID",
+			topic:    fmt.Sprintf("%s.%s", topic, subtopic),
+			clientID: "clientid7",
+			err:      nil,
+			handler:  handler{false},
+		},
+		{
+			desc:     "Subscribe to an empty topic with an ID",
+			topic:    "",
+			clientID: "clientid7",
+			err:      mqtt_pubsub.ErrEmptyTopic,
+			handler:  handler{false},
+		},
+		{
+			desc:     "Subscribe to a topic with empty id",
+			topic:    topic,
+			clientID: "",
+			err:      mqtt_pubsub.ErrEmptyID,
+			handler:  handler{false},
+		},
+	}
+	for _, tc := range cases {
+		err := pubsub.Subscribe(tc.clientID, tc.topic, tc.handler)
+		switch err {
+		case nil:
+			// if no error, publish message, and receive after subscribing
+			expectedMsg := messaging.Message{
+				Channel:  channel,
+				Subtopic: subtopic,
+				Payload:  data,
+			}
+			token := client.Publish(tc.topic, qos, false, expectedMsg.Payload)
+			token.Wait()
+
+			receivedMsg := <-msgChan
+			assert.Equal(t, expectedMsg.Payload, receivedMsg.Payload, fmt.Sprintf("%s: expected %+v got %+v\n", tc.desc, expectedMsg, receivedMsg))
+		default:
+			switch tc.err {
+			case nil:
+				assert.Nil(t, err, "%s: got unexpected error: %s", tc.desc, err.Error())
+			default:
+				assert.Equal(t, err, tc.err, fmt.Sprintf("%s: expected: %s, but got: %s", tc.desc, err, tc.err))
+			}
+		}
+	}
+}
+
+// Tests both Subscribe and Unsubscribe
```

Comment:
No need for this test. These cases are part of either `TestSubscribe` or, more likely, `TestUnsubscribe`.

Reply:
The unsubscribe test uses client.subscribe, while testsubunsub uses pubsub to subscribe and unsubscribe, so it tests the complete cycle of sub and unsub.
I created this, since there is also a testpubsub, which tests pubsub.publish and pubsub.subscribe together

Reply:
Makes sense. Please rename it to `TestSubscriber` since you are testing - the `Subscriber` interface. Add some explanation in the comment explaining how this differs from `TestSubscribe` and `TestUnsubscribe`.